packages/remix-react/components.tsx

```diff
@@ -1296,7 +1296,45 @@ export function useMatches() {
  *
  * @see https://remix.run/api/remix#useloaderdata
  */
-export function useLoaderData<T = AppData>(): T {
+
+export type TypedResponse<T> = Response & {
+  json(): Promise<T>;
+};
+
+type DataFunction = (...args: any[]) => any; // matches any function
+type DataOrFunction = AppData | DataFunction;
+type JsonPrimitives = string | number | boolean | null;
+type NonJsonPrimitives = undefined | Function | symbol;
+type SerializeType<T> = T extends JsonPrimitives
+  ? T
+  : T extends undefined
+  ? undefined
+  : T extends { toJSON(): infer U }
+  ? U
+  : T extends []
+  ? []
+  : T extends [any, ...any[]]
```

Comment:
Could we use `unknown` instead?
https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html#any

```suggestion
  : T extends [unknown, ...unknown[]]
```